File: NetDeck/Faction.m
//
//  Faction.m
//  Net Deck
//
//  Created by Gereon Steffens on 15.12.13.
//  Copyright (c) 2014 Gereon Steffens. All rights reserved.
//

#import "Faction.h"
#import "CardManager.h"
#import "TableData.h"
#import "SettingsKeys.h"

@implementation Faction

static NSMutableDictionary* faction2name;

static NSMutableArray* runnerFactions;
static NSMutableArray* runnerFactionsPreDAD;
static NSMutableArray* corpFactions;
static NSDictionary* code2faction;
static TableData* allFactions;

+(void) initialize
{
    faction2name = [NSMutableDictionary dictionary];
    faction2name[@(NRFactionNone)] = kANY;
    
    code2faction = @{
                     @"anarch": @(NRFactionAnarch),
                     @"shaper": @(NRFactionShaper),
                     @"criminal": @(NRFactionCriminal),
                     @"weyland-consortium": @(NRFactionWeyland),
                     @"haas-bioroid": @(NRFactionHaasBioroid),
                     @"nbn": @(NRFactionNBN),
                     @"jinteki": @(NRFactionJinteki),
                     @"adam": @(NRFactionAdam),
                     @"apex": @(NRFactionApex),
                     @"sunny-lebeau": @(NRFactionSunnyLebeau),
                     @"neutral": @(NRFactionNeutral)
                     };
}

+(void) initializeFactionNames:(NSArray*)cards
{
    [faction2name removeAllObjects];
    faction2name[@(NRFactionNone)] = kANY;
    
    for (Card* c in cards)
    {
        [faction2name setObject:c.factionStr forKey:@(c.faction)];
    }
    [faction2name setObject:@"Adam" forKey:@(NRFactionAdam)];
    [faction2name setObject:@"Apex" forKey:@(NRFactionApex)];
    [faction2name setObject:@"Sunny Lebeau" forKey:@(NRFactionSunnyLebeau)];
    
    NRFaction rf_all[] = { NRFactionAnarch, NRFactionCriminal, NRFactionShaper, NRFactionAdam, NRFactionApex, NRFactionSunnyLebeau };
    NRFaction rf_preDaD[] = { NRFactionAnarch, NRFactionCriminal, NRFactionShaper };
    NRFaction cf[] = { NRFactionHaasBioroid, NRFactionJinteki, NRFactionNBN, NRFactionWeyland };
    NSArray* common = @[ [Faction name:NRFactionNone ], [Faction name:NRFactionNeutral ]];
    
    runnerFactions = [NSMutableArray array];
    for (int i=0; i<DIM(rf_all); ++i)
    {
        [runnerFactions addObject:[Faction name:rf_all[i]]];
    }
    
    runnerFactionsPreDAD = [NSMutableArray array];
    for (int i=0; i<DIM(rf_preDaD); ++i)
    {
        [runnerFactionsPreDAD addObject:[Faction name:rf_preDaD[i]]];
    }
    
    corpFactions = [NSMutableArray array];
    for (int i=0; i<DIM(cf); ++i)
    {
        [corpFactions addObject:[Faction name:cf[i]]];
    }

    NSArray* factionSections = @[ @"", l10n(@"Runner"), l10n(@"Corp") ];
    // NB copy is important, both faction arrays are modified below
    NSArray* factions = @[ common, runnerFactions.copy, corpFactions.copy ];
    allFactions = [[TableData alloc] initWithSections:factionSections andValues:factions];
    
    NSMutableIndexSet* indexes = [NSMutableIndexSet indexSetWithIndex:0];
    [indexes addIndex:1];
    [runnerFactions insertObjects:common atIndexes:indexes];
    [runnerFactionsPreDAD insertObjects:common atIndexes:indexes];
    [corpFactions insertObjects:common atIndexes:indexes];
}

+(NSString*) shortName:(NRFaction)faction
{
    switch (faction)
    {
        case NRFactionHaasBioroid:
            return @"H-B";
        case NRFactionWeyland:
            return @"Weyland";
        case NRFactionSunnyLebeau:
            return @"Sunny";
        default:
            return [Faction name:faction];
    }
}

+(NSString*) name:(NRFaction)faction
{
    return [faction2name objectForKey:@(faction)];
}

+(NRFaction) faction:(NSString*)faction
{
    NSNumber* f = [code2faction objectForKey:[faction lowercaseString]];
    return f == nil ? NRFactionNone : (NRFaction)f.integerValue;
}

+(NSArray*) factionsForRole:(NRRole)role
{
    NSAssert(role != NRRoleNone, @"no role");
    BOOL dataDestinyAllowed = [[NSUserDefaults standardUserDefaults] boolForKey:USE_DATA_DESTINY];
    
    if (role == NRRoleRunner)
    {
        return dataDestinyAllowed ? runnerFactions : runnerFactionsPreDAD;
    }
    return corpFactions;
}

+(TableData*) allFactions
{
    return allFactions;
}

@end

File: NetDeck/Faction.swift
//
//  Faction.swift
//  NetDeck
//
//  Created by Gereon Steffens on 14.11.15.
//  Copyright © 2021 Gereon Steffens. All rights reserved.
//

import Foundation
import UIKit.UIColor
import SwiftyUserDefaults

// @objc to make NSPredicates on card.faction work

@objc enum Faction: Int {
    
    case none = -1
    case neutral
    
    case haasBioroid, weyland, nbn, jinteki
    
    case anarch, shaper, criminal
    
    case adam, apex, sunnyLebeau

    private static var faction2name = [Faction: String]()
    
    static let runnerFactionsCore: [Faction] = [ .anarch, .criminal, .shaper ]
    static let runnerMiniFactions: [Faction] = [ .adam, .apex, .sunnyLebeau ]
    static let runnerFactionsAll = runnerFactionsCore + runnerMiniFactions
    
    private(set) static var runnerFactionNamesAll = [String]()
    private(set) static var runnerFactionNamesCore = [String]()
    
    static let corpFactions: [Faction] = [ .haasBioroid, .jinteki, .nbn, .weyland ]
    private(set) static var corpFactionNames = [String]()
    
    private static var allFactions: TableData<String>!
    private static var allFactionsCore: TableData<String>!
    
    static let weylandConsortium = "Weyland Consortium"
    
    static func name(for faction: Faction) -> String {
        return Faction.faction2name[faction] ?? "n/a"
    }
    
    static func initializeFactionNames(_ factionsDict: [Faction: String]) -> Bool {
        faction2name = [:]
        faction2name[.none] = Constant.kANY
        faction2name[.neutral] = "Neutral".localized()
        
        let expectedNames = runnerFactionsAll.count + corpFactions.count + 2 // +2 for "any" and "neutral"
        faction2name.merge(factionsDict) { (current, _) in return current }
        assert(faction2name.count == expectedNames)
        if faction2name.count != expectedNames {
            return false
        }

        runnerFactionNamesAll = []
        runnerFactionNamesCore = []
        corpFactionNames = []

        let common = [ Faction.name(for: .none), Faction.name(for: .neutral) ]
        
        for faction in runnerFactionsAll {
            runnerFactionNamesAll.append(Faction.name(for: faction))
        }
        for faction in runnerFactionsCore {
            runnerFactionNamesCore.append(Faction.name(for: faction))
        }
        for faction in corpFactions {
            corpFactionNames.append(Faction.name(for: faction))
        }
        
        let factionSections = [ "", "Runner".localized(), "Corp".localized() ]
        let factionsAll = [ common, runnerFactionNamesAll, corpFactionNames ]
        let factionsCore = [ common, runnerFactionNamesCore, corpFactionNames ]

        allFactions = TableData(sections: factionSections, values: factionsAll)
        allFactionsCore = TableData(sections: factionSections, values: factionsCore)
        
        runnerFactionNamesAll.insert(contentsOf: common, at: 0)
        runnerFactionNamesCore.insert(contentsOf: common, at: 0)
        corpFactionNames.insert(contentsOf: common, at: 0)
        
        return true
    }
    
    static func factionsFor(role: Role, packUsage: PackUsage) -> [String] {
        assert(role != .none, "no role")
        
        if role == .runner {
            if packUsage == .selected {
                let dataDestinyAllowed = Defaults[.useDataDestiny]
                return dataDestinyAllowed ? runnerFactionNamesAll : runnerFactionNamesCore
            } else {
                return runnerFactionNamesAll
            }
        } else {
            return corpFactionNames
        }
    }
    
    static func factionsForBrowser(packUsage: PackUsage) -> TableData<String> {
        if packUsage == .selected {
            let dataDestinyAllowed = Defaults[.useDataDestiny]
            return dataDestinyAllowed ? allFactions : allFactionsCore
        } else {
            return allFactions
        }
    }
    
    static func shortName(for faction: Faction) -> String {
        switch faction {
        case .haasBioroid: return "H-B"
        case .weyland: return "Weyland".localized()
        case .sunnyLebeau: return "Sunny"
        default: return Faction.name(for: faction)
        }
    }
    
    // needed for the jinteki.net uploader
    static func fullName(for faction: Faction) -> String {
        switch faction {
        case .weyland:
            return Faction.weylandConsortium
        default:
            return Faction.name(for: faction)
        }
    }
}

extension Faction {
    private static let colors: [Faction: UInt] = [
        .jinteki:      0x940c00,
        .nbn:          0xd7a32d,
        .weyland:      0x2d7868,
        .haasBioroid:  0x6b2b8a,
        .shaper:       0x6ab545,
        .criminal:     0x4f67b0,
        .anarch:       0xf47c28,
        .adam:         0xae9543,
        .apex:         0xa8403d,
        .sunnyLebeau:  0x776e6f
    ]

    static func color(for faction: Faction) -> UIColor {
        guard let hex = hexColor(for: faction) else {
            return .label
        }

        return UIColor(rgb: hex)
    }
    
    static func hexColor(for faction: Faction) -> UInt? {
        return Faction.colors[faction]
    }
    
}

